common/lessonplan/generate.ts

```diff
@@ -0,0 +1,298 @@
+                        loader = new DocxLoader(blob);
+                        docs = await loader.load();
+                    } else if (file.mimetype === 'application/vnd.openxmlformats-officedocument.presentationml.presentation') {
+                        loader = new PPTXLoader(blob);
+                        docs = await loader.load();
+                    } else {
+                        throw new Error(`Unsupported file type: ${file.mimetype}`);
+                    }
+
+                    // Add filename and page numbers (only for PDFs) to the content
+                    const content =
+                        `Filename: ${file.originalname}\n\n` +
+                        docs
+                            .map((doc, index) => (file.mimetype === 'application/pdf' ? `Page ${index + 1}:\n${doc.pageContent}` : doc.pageContent))
+                            .join('\n\n');
+
+                    // Check if the content is empty (excluding filename and page numbers)
+                    const contentWithoutMetadata = content
+                        .replace(`Filename: ${file.originalname}\n\n`, '')
+                        .replace(/Page \d+:\n/g, '')
+                        .trim();
+                    if (contentWithoutMetadata.length === 0) {
+                        emptyFiles.push(file.originalname);
+                        return null;
+                    }
+
+                    return {
+                        name: file.originalname,
+                        content: content,
+                        type: file.mimetype,
+                    };
+                } catch (error) {
+                    logger.error(`Error fetching or parsing file with UUID ${uuid}: ${error.message}`);
+                    throw error;
+                }
+            })
+        );
+
+        const validFiles = allFiles.filter((file): file is NonNullable<typeof file> => file !== null);
+
+        if (emptyFiles.length > 0) {
+            throw new Error(`The following files are empty or their content could not be processed: ${emptyFiles.join(', ')}`);
+        }
+
+        // Combine all file contents
+        combinedContent = validFiles.map((file) => file.content).join('\n\n');
+    }
+
+    // Check for OpenAI API key
+    if (!process.env.OPENAI_API_KEY) {
+        throw new Error('OpenAI API key is not set. Please set the OPENAI_API_KEY environment variable.');
+    }
+
+    // Initialize the ChatOpenAI model
+    const model = new ChatOpenAI({
+        modelName: MODEL_NAME,
+        temperature: 0,
+    });
+
+    // Create a new plan schema based on expected outputs
+    const dynamicPlan = z.object(Object.fromEntries(Object.entries(plan.shape).filter(([key]) => !expectedOutputs || expectedOutputs.includes(key))));
+
+    const structuredLlm = model.withStructuredOutput(dynamicPlan);
+
+    try {
+        logger.info('Generating lesson plan...');
```

Comment:
This log could use some context to be useful, e.g. `, { dynamicPlan, prompt, ... }`